odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1516 @@
+    functionals are defined on a Hilbert space, i.e., that we consider
+    functionals
+
+    .. math::
+
+            f : X \\to F,
+
+    where :math:`X` is a Hilbert space and :math:`F` is a field of scalars
+    associated with :math:`X`. This has been done in order to simplify the
+    concept of *convex conjugate functional*. Since Hilbert spaces as selfdual
+    the convex conjugate functional is defined as
+
+    .. math::
+
+            f^* : X \\to F,
+
+            f^*(x^*) = \\sup_{x} \{ \\langle x^*,x \\rangle - f(x)  \}.
+
+    See, e.g., [Lue1969]_, [Roc1970]_ and [BC2011]_.
+    """
+
+    # TODO: Update doc above. What to write?
+
+    def __init__(self, domain, linear=False, smooth=False, concave=False,
+                 convex=False, grad_lipschitz=np.inf):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        domain : `Set`
+            The domain of this functional, i.e., the set of elements to
+            which this functional can be applied
+        linear : `bool`
+            If `True`, the functional is considered as linear. In this
+            case, ``domain`` and ``range`` have to be instances of
+            `LinearSpace`, or `Field`.
+        smooth : `bool`, optional
+            If `True`, assume that the functional is continuously
+            differentiable
+        concave : `bool`, optional
+            If `True`, assume that the functional is concave
+        convex : `bool`, optional
+            If `True`, assume that the functional is convex
+        grad_lipschitz : 'float', optional
+            The Lipschitz constant of the gradient.
+        """
+
+        Operator.__init__(self, domain=domain,
+                          range=domain.field, linear=linear)
+        self._is_smooth = bool(smooth)
+        self._is_convex = bool(convex)
+        self._is_concave = bool(concave)
+        self._grad_lipschitz = float(grad_lipschitz)
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional.
+
+        Notes
+        -----
+        The operator that corresponds to the mapping
+
+        .. math::
+
+            x \\to \\nabla f(x)
+
+        where :math:`\\nabla f(x)` is the element used to evaluated
+        derivatives in a direction :math:`d` by
+        :math:`\\langle \\nabla f(x), d \\rangle`.
+        """
+        raise NotImplementedError
+
+    def proximal(self, sigma=1.0):
+        """Return the proximal operator of the functional.
+
+        Parameters
+        ----------
+        sigma : positive float, optional
+            Regularization parameter of the proximal operator.
+
+        Returns
+        -------
+        out : `Operator`
+            Domain and range equal to domain of functional.
+
+        Notes
+        -----
+        Note that ``Functional.proximal`` is a `Proximal factory`.
+        """
+        raise NotImplementedError
+
+    @property
+    def conjugate_functional(self):
```

Comment:
We use cconj in all of the other code (solvers and proximals). So that would be my goto naming. If we change here we should change everywhere.